core/src/main/java/io/grpc/internal/ServiceConfigUtil.java

```diff
@@ -642,4 +697,64 @@ private static long saturatedAdd(long a, long b) {
     // we did over/under flow, if the sign is negative we should return MAX otherwise MIN
     return Long.MAX_VALUE + ((naiveSum >>> (Long.SIZE - 1)) ^ 1);
   }
+
+  /**
+   * A LoadBalancingConfig that includes the policy name (the key) and its raw config value (parsed
+   * JSON).
+   */
+  public static final class LbConfig {
+    private final String policyName;
+    private final Map<String, Object> rawConfigValue;
+
+    public LbConfig(String policyName, Map<String, Object> rawConfigValue) {
+      this.policyName = checkNotNull(policyName, "policyName");
+      this.rawConfigValue = checkNotNull(rawConfigValue, "rawConfigValue");
+    }
+
+    public String getPolicyName() {
+      return policyName;
+    }
+
+    public Map<String, Object> getRawConfigValue() {
+      return rawConfigValue;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o instanceof LbConfig) {
+        LbConfig other = (LbConfig) o;
+        return policyName.equals(other.policyName)
+            && rawConfigValue.equals(other.rawConfigValue);
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(policyName, rawConfigValue);
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("policyName", policyName)
+          .add("rawConfigValue", rawConfigValue)
+          .toString();
+    }
+  }
+
+  /**
+   * Thrown when the config is a valid JSON, but doesn't conform to the config scheme.
+   */
+  public static final class MalformedConfigException extends Exception {
+    private static final long serialVersionUID = 1L;
+
+    private MalformedConfigException(String msg) {
+      super(msg);
+    }
+
+    private MalformedConfigException(String msg, Throwable cause) {
+      super(msg, cause);
+    }
+  }
```

Comment:
ServiceConfigUtil doesn't seem like an appropriate name any more.  I think you should rename the class, or move the inner types out.

Reply:
I don't want `ServiceConfigUtil` to be public. Maybe a particular method/class may need to be split out, but there's far too many random things going on here. For example, things like getFallbackPolicyFromXdsConfig should really live in XDS.

Reply:
>getFallbackPolicyFromXdsConfig should really live in XDS.

Yeah, I was considering putting `getFallbackPolicyFromXdsConfig` in XDS, but it reuses too many methods in `ServiceConfigUtil`, and need large refactoring so I gave up.